### ISSUE TITLE
Support of generic odata client

### ISSUE THREAD

**Issue by trepeschr** [SAP/cloud-sdk-js#1089]
Dear Cloud-SDK-Team, 
First of all we want to thank you for the good work you put up with the Cloud SDK for JavaScript. 
We are currently testing the JS-version of the cloud SDK in our company for different purposes and we like the features it brings especially for odata and as a software developer I highly appreciate the clean architecture with the monorepo-approach with lerna, the usage of an architecture decision record (knowledge-base/adr) and the appropriate usage of typescript in this project. 

Right now, we have one requested feature, 
that would bring a big benefit to us as developers and would accelerate the use within our company: The Generic Odata Client feature.  Unlike the JAVA SDK, the cloud sdk for JS needs statically typed, dynamically generated typescript files as an internal dependency. According to the following link [feature matrix](https://sap.github.io/cloud-sdk/docs/overview/cloud-sdk-feature-matrix/), the Cloud SDK JS-version is not supporting the dynamic generation of these "artifacts" or a dynamic odata client like in the JAVA version.

Within our team, we are seriously considering / evaluating the future use of the SDK  even for SAP on-premise-landscapes, 
where we would like to use the generators of the Cloud SDK for building and filtering API-requests, 
as they are really nice and easy to use and it would bring us much more consistency and simplicity over different plattforms as a benefit. However, before we can use it, we have to answer some questions. 

1. Is it allowed by SAP to use parts of the Cloud SDK in this SAP On-Premise-scenarios?
  
2. Is the generic odata client on the roadmap of the product for this year? 

3. In our purely technical evaluation, we found out, that there would be a technical issue, 
that would prevent us from using the sdk for the aforementioned scenario: 
Once we deploy the generated typescript-files and some fields of the ODATA-Service on the Netweaver-backend are changing,
we experience the issue, that we would have to rebuild or re-deploy the generated typescript-files each time, the backend-service or metadata changes. In that usecase a more dynamic way without requiring the direct use of the file system would help us a lot. 

Thank you in advance. 

Best Regards
Roland Trepesch

**Comment by marikaner** [comment]
Hey @trepeschr, thanks a lot for the warm words, it's still early, but it might make my day ;)

1. I don't see any issue with that. I assume you are asking this for legal reasons? If you are generating clients for SAP services, you can use those clients, but you may not publish them. The SAP S/4HANA services (Cloud and On-Premise) are SAP's intellectual property and their use is restricted. I don't know what the regulations are if you want to publish them internally.
Are you aware of our [pre-generated OData clients for OP](https://sap.github.io/cloud-sdk/docs/js/release-notes-sap-cloud-sdk-for-javascript-and-typescript#new-functionality) (I noticed that we might improve the documentation regarding this. Noted.) and in case not would that fulfil your requirements? They are available under the SAP Developer license.
2. Currently it is not. We have investigated this idea last year and decided to postpone this until somebody (else) asks for it. Well here we are ;)
The reason why we postponed this is that our architecture currently does not allow to achieve this easily, if we want to reuse it (which we do). Can you elaborate a little on your use case and whether you have any special requirements towards this?
I think it makes sense to collect your requirements in a separate ticket (feature request) and create an ADR based on this. We might consider this for our roadmap, but in case we can't with an ADR it might be contributed by someone.
3. We have talked about this as well in the past. If I understand correctly you would like to have a generic OData client, that is typed "on-the-fly"?
How often are the services changing and why is it a problem to regenerate and republish the clients? It would be helpful to understand the background of this a little better to put this into perspective.

If you want we could also setup a call to discuss details.

**Comment by FrankEssenberger** [comment]
Hello @trepeschr ,

regarding the point of changing backend services. Please note that we have the [custom fields](https://sap.github.io/cloud-sdk/docs/js/features/odata/use-odata-v2-type-safe-client-for-javascript-typescript/#custom-fields) "bucket" in which all data which is not part of the static type should end up. An in JS you can handle object very dynamically. Perhaps this could solve the issue on frequently changing service specs.

Best
Frank

**Comment by trepeschr** [comment]
Hello @FrankEssenberger , @marikaner. 
Thank you for your super-fast and insightful answers, that are already helping me at least a bit. 
I was not aware of the "custom fields"-option and we tried it out and it indeed works for Custom-Services. 

@marikaner: Regarding your question: _We have talked about this as well in the past. 
If I understand correctly you would like to have a generic OData client, that is typed "on-the-fly"?_ : 
Basically, the generic client was what we were looking for, but as there is a workaround, it's completely fine for use, to stick with the statically typed client for now.

**Comment by FrankEssenberger** [comment]
@trepeschr is it ok if we close this issue?

**Comment by trepeschr** [comment]
Yes, that's fine for me.

**Comment by artemkovalyov** [comment]
Thanks for the feedback @trepeschr. I'm happy we have found a working solution thanks to @marikaner and @FrankEssenberger. I'll be closing this for the time being. We're happy to see you back with questions and feedback whenever you feel smth can be improved. Cheers!
